modules/programs/hexchat.nix

```diff
@@ -0,0 +1,256 @@
+{ config, pkgs, lib, ... }:
+with builtins;
+with lib;
+let cfg = config.programs.hexchat;
+in {
+  meta.maintainers = [ maintainers.superherointj ];
+
+  options.programs.hexchat = {
+    channels = with lib.types;
+      let
+        modChannelOption = submodule {
+          options = {
+            options = let
+              channelOptions = submodule {
+                options = {
+                  autoconnect = mkOption {
+                    type = nullOr bool;
+                    description = "Autoconnect to network";
+                    default = false;
+                  };
+                  connect_to_selected_server_only = mkOption {
+                    type = nullOr bool;
+                    description = "Connect to selected server only";
+                    default = true;
+                  };
+                  bypass_proxy = mkOption {
+                    type = nullOr bool;
+                    description = "Bypass proxy";
+                    default = true;
+                  };
+                  use_ssl_for_all_servers = mkOption {
+                    type = nullOr bool;
+                    description = "Use SSL for all servers";
+                    default = false;
+                  };
+                  accept_invalid_ssl_certificates = mkOption {
+                    type = nullOr bool;
+                    description = "Accept invalid SSL certificates";
+                    default = false;
+                  };
+                  use_global_user_information = mkOption {
+                    type = nullOr bool;
+                    description = "Use global user information";
+                    default = false;
+                  };
+                };
+              };
```

Comment:
I think it would be more readable to move this submodule definition to the `let` outside of the module, the same one of the line 4.

Reply:
My guess for the lack of readability stem from:
1) Having options fields nested. As a solution I have alphabetically ordered field names. (Looks better to me now. Let me know.)
2) nixfmt does not help much in readability.
---
If I were to move to line 4, it would make it more confusing because that field type is only used in that field, and whoever is reading that line/field will need/want to understand the field type too, and it is at the right scope as nothing else will depend on it.
---
I have made some changes. Let me know if it does look better for you.

Reply:
Better but still difficult to grasp. `let` scooped for only one attrset looks better when the let is small (like 1-5 lines), but this is huge so by the time you finish reading the contents of `let` you already loosed context of what you're reading.

> If I were to move to line 4, it would make it more confusing because that field type is only used in that field, it is at the right scope as nothing else will depend on it and whoever is reading that line/field will need/want to understand the field type too.

I really don't think so, this is a pretty common pattern for bigger `let` so I don't think anyone would have difficult with it. So I still in favor of moving the definition of `modChannelOption` to the `let` in line 4.

Reply:
I have looked at code again and been thinking about this. I'm not sure this change would be an improvement. Because "channelOptions" is limited to the field 'options' and has no reuse and it is already at the right scope. 'channelOptions' code isn't that large even. Moving it to a lower layer would just add noise to the wrong layer as this code is specific to this upper layer. The tradeoff of lowering layers, messing with scope and adding noise to lower layers doesn't seem meaningful to me. Can we just leave it as it is?

Reply:
> Moving it to a lower layer would just add noise to the wrong layer as this code is specific to this upper layer. The tradeoff of lowering layers, messing with scope and adding noise to lower layers doesn't seem meaningful to me. Can we just leave it as it is?

It is not only to reduce the number of indentations, "when in Rome do what Romans do". Even modules in nixpkgs (where there is no nixfmt) generally define submodules at the highest level, even if used only once. IMO, it will actually reduce the noise since like you said, less indentations.